File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/screens/CountryPickerSheet.kt
package com.usesmileid.sampleapps.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import com.smileid.designsystem.SmileDimens
import com.usesmileid.sampleapps.ui.UseSmileIDSampleTestIds
import com.usesmileid.sampleapps.ui.components.UseSmileIDSampleEmptyState
import com.usesmileid.sampleapps.ui.components.UseSmileIDSampleFullHeightBottomSheet
import com.usesmileid.sampleapps.ui.components.UseSmileIDSampleOptionRow
import com.usesmileid.sampleapps.ui.components.UseSmileIDSampleSearchField
import com.usesmileid.sampleapps.ui.state.UseSmileIDSampleCountry
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/** The country picker. A full-height sheet, because the list is long enough that a partial one fights the keyboard. */
@Composable
fun CountryPickerSheet(
    selected: UseSmileIDSampleCountry?,
    query: String,
    onQueryChange: (String) -> Unit,
    onSelect: (UseSmileIDSampleCountry) -> Unit,
    onDismissRequest: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val matches = UseSmileIDSampleCountry.entries.filter { it.label.contains(query, ignoreCase = true) }
    UseSmileIDSampleFullHeightBottomSheet(
        title = "Country",
        onDismissRequest = onDismissRequest,
        modifier = modifier,
        testId = UseSmileIDSampleTestIds.COUNTRY_SHEET,
    ) {
        UseSmileIDSampleSearchField(
            query = query,
            onQueryChange = onQueryChange,
            placeholder = "Search country",
            testId = UseSmileIDSampleTestIds.COUNTRY_SEARCH,
        )
        PickerList(
            empty = matches.isEmpty(),
            emptyLabel = "No country matches \u201c$query\u201d",
            emptyTestId = UseSmileIDSampleTestIds.COUNTRY_EMPTY,
        ) {
            matches.forEach { country ->
                UseSmileIDSampleOptionRow(
                    label = country.label,
                    selected = country == selected,
                    onClick = { onSelect(country) },
                    leadingText = country.flag,
                    testId = UseSmileIDSampleTestIds.countryOption(country.code),
                )
            }
        }
    }
}

/** An empty result is a state a search must have, or a typo looks like a broken sheet. */
@Composable
internal fun PickerList(
    empty: Boolean,
    emptyLabel: String,
    emptyTestId: String,
    content: @Composable () -> Unit,
) {
    if (empty) {
        UseSmileIDSampleEmptyState(text = emptyLabel, testId = emptyTestId)
        return
    }
    Column(
        modifier = Modifier.verticalScroll(rememberScrollState()),
        verticalArrangement = Arrangement.spacedBy(SmileDimens.spacingXxs),
    ) {
        content()
    }
}

File: ios/SampleUI/Sources/SampleUI/Screens/CountryPickerSheet.swift
import SwiftUI

/// The country picker, presented as a layer over the form that owns it rather than a destination.
public struct CountryPickerSheet: View {
  private let selected: UseSmileIDSampleCountry?
  @Binding private var query: String
  private let onSelect: (UseSmileIDSampleCountry) -> Void

  public init(
    selected: UseSmileIDSampleCountry?,
    query: Binding<String>,
    onSelect: @escaping (UseSmileIDSampleCountry) -> Void
  ) {
    self.selected = selected
    _query = query
    self.onSelect = onSelect
  }

  public var body: some View {
    UseSmileIDSampleBottomSheet(title: "Country", testId: UseSmileIDSampleTestIds.countrySheet) {
      UseSmileIDSampleSearchField(
        query: $query,
        placeholder: "Search country",
        testId: UseSmileIDSampleTestIds.countrySearch
      )
      UseSmileIDSamplePickerList(
        isEmpty: matches.isEmpty,
        emptyLabel: "No country matches \u{201C}\(query)\u{201D}",
        emptyTestId: UseSmileIDSampleTestIds.countryEmpty
      ) {
        ForEach(matches, id: \.self) { country in
          UseSmileIDSampleOptionRow(
            label: country.label,
            leadingText: country.flag,
            selected: country == selected,
            testId: UseSmileIDSampleTestIds.countryOption(country.code),
            onTap: { onSelect(country) }
          )
        }
      }
    }
  }

  private var matches: [UseSmileIDSampleCountry] {
    UseSmileIDSampleCountry.matching(query)
  }
}

/// An empty result is a state a search must have, or a typo looks like a broken sheet. Internal and
/// beside its first caller, as the Compose twin is: the two picker sheets are its only consumers.
struct UseSmileIDSamplePickerList<Content: View>: View {
  private let isEmpty: Bool
  private let emptyLabel: String
  private let emptyTestId: String
  private let content: Content

  init(
    isEmpty: Bool,
    emptyLabel: String,
    emptyTestId: String,
    @ViewBuilder content: () -> Content
  ) {
    self.isEmpty = isEmpty
    self.emptyLabel = emptyLabel
    self.emptyTestId = emptyTestId
    self.content = content()
  }

  var body: some View {
    if isEmpty {
      UseSmileIDSampleEmptyState(text: emptyLabel, testId: emptyTestId)
    } else {
      VStack(spacing: SmileSpacing.spacingXxs) { content }
    }
  }
}
